### DISCUSSION TITLE
Future Passkey Support

### DISCUSSION THREAD

**Discussion by coolrazor007** [dani-garcia/vaultwarden#3355] [General]
This is probably way to preemptive, but I was wondering about passkey support when the Bitwarden team eventually includes it on their side.  I've been getting more excited about passkeys/FIDO2.

I guess my question (not a fair one) is how long after passkeys are supported by Bitwarden will they get implemented in Vaultwarden?  Or just any thoughts about passkeys from those involved with Vaultwarden.

**Comment by RealOrangeOne** [comment]
> how long after passkeys are supported by Bitwarden will they get implemented in Vaultwarden

When someone has the time to write the change, plus however long it takes to write the change based on complexity.

**Comment by BlackDex** [comment]
If you are revering to passwordless login i suggest to read this post: https://github.com/dani-garcia/vaultwarden/discussions/2817
And besides that, that feature isn't even available on the Bitwarden Self-Hosted instances, so it would be difficult for us to implement this also.

**Comment by coolrazor007** [comment]
Here's the Bitwarden feature request I'm referring to: https://community.bitwarden.com/t/store-webauthn-fido2-credentials-in-bitwarden-passkey-support/42153

As someone who isn't a Rust dev, how can I help once development can begin?  (I understand that this is far out)

**Reply by BlackDex** [reply]
First this functionality should be provided for self-hosted, after that we can check and see if we can add this or not.
In the end, if and when this feature can be developed for Vaultwarden, then a way to help would be help testing this.

**Comment by lcheylus** [comment]
Bitwarden announced recently **Passkey support** : https://bitwarden.com/blog/bitwarden-passkey-management/
- Store passkeys in Bitwarden
- Sign in and unlock Bitwarden with passkeys

Is it planned to support Passkey in Vaultwarden ?

**Reply by BlackDex** [reply]
First it needs to be released, then we can see if that is possible.
Storing the keys is probably not a big issue, new table to the database, some endpoints.

But allowing login might be different. But, i have no clue yet.

**Reply by Schluggi** [reply]
Any updates? 😊

**Reply by BlackDex** [reply]
Well Vaultwarden already supports it via #3593 , but the clients not yet, unless you build them your self.

**Reply by Schluggi** [reply]
Thank you for the update. "Client" means browser extension?

**Reply by BlackDex** [reply]
Yes

**Reply by oguzhane** [reply]
@BlackDex #3593 appears to add support of storing passkeys.

Does it support sign-in and unlock Bitwarden with passkey though?

**Reply by BlackDex** [reply]
Not sure actually how passkey works, but if that is based upon WebAuthn, then it should probably work.

But, i have not tested it nor looked at the code.

**Reply by oguzhane** [reply]
Perhaps @GeekCornerGH can shed a light on this.

**Reply by megamorf** [reply]
Passkeys are based on asymmetric encryption where the private key remains on the user/device side, in this case, the user's vault. As explained already, while the implementation on the server side has been worked on, the release of updated Bitwarden clients that can leverage the feature is still pending. So we'll have to wait until the different Bitwarden clients get a new release with support for passkeys.

**Reply by oguzhane** [reply]
@megamorf thanks for highlighting the concept.
My understanding is that the pr brings the ability for external websites where the private key stored in user's vault. i was actually wondering ability to sign-in and unluck bitwarden(i.e. webui) itself using passkeys.

i guess in that case private key(that is used to unlock bitwarden) should be stored in out of the vault, on user's side such as yubikey etc? Hence, if this is the case, the pr doesn't aim to bring that feature.

**Reply by megamorf** [reply]
FYI that blog post was recently updated with the following note:

> Editor's note September 5, 2023: Passkey storage in Bitwarden Password Manager will be released in October. Sign in with passkey will come shortly after.

**Reply by danieletorelli** [reply]
This has been merged today: https://github.com/bitwarden/server/pull/2679 🎉

**Reply by iwa** [reply]
they planned an [event on November 9](https://bitwarden.com/events/passkeys-and-bitwarden/) dedicated to passkeys, so it's coming really soon:tm:

**Reply by leikoilja** [reply]
aaaand, it's out :)

**Reply by AlexCherrypi** [reply]
Does login to Vaultwarden via Passkeys (no masterkey) work now?

**Comment by danieletorelli** [comment]
And it's finally merged: https://github.com/bitwarden/server/pull/2679 🎉

**Reply by coolrazor007** [reply]
Awesome!

**Reply by idoodler** [reply]
Now, what does that mean? I guess it has to be implemented in Vaultwarden first and also the Clients need to support Passkeys before we can use it, right?

**Reply by danieletorelli** [reply]
That's correct. This means though that is very close to be officially released so the clients will support it soon, if they don't already.

**Comment by ilbarone87** [comment]
Most of them PRs for passkeys have been merged also in the client repo https://github.com/bitwarden/clients/pull/4715 🎉🎉🎉
Really close to the release

**Comment by afonsofrancof** [comment]
Can't wait for this to be released.
I am tired of typing my master password to unlock the vault and using passkeys on apple password chain to login to websites.
Much thanks to everyone involved.

**Comment by ItsShadowCone** [comment]
@danieletorelli did you have the time to look at the upstream PRs, can you estimate how much work needs to be done on the vaultwarden side? Referencing #3593, it seems that implementation was a little premature but laid most of the groundwork?

**Reply by danieletorelli** [reply]
Unfortunately I don't have a way to estimate the complexity because I'm unfamiliar with both the codebases :(
Hopefully some contributor will be able to answer here.

**Comment by ilbarone87** [comment]
Clients and Server for bitwarden upstream have been just released!!!

**Comment by likevi9528** [comment]
[bitwarden/server](https://github.com/bitwarden/server/releases/tag/v2023.10.0) : 
...
Add support for storing passkeys on login items
...

[bitwarden/clients](https://github.com/bitwarden/clients/releases/tag/web-v2023.10.0)  : 
...
View and edit login items with passkeys saved to them
...

**Comment by BitesizedLion** [comment]
https://bitwarden.com/help/releasenotes/#2023-10-0

> Save passkeys to your vault: Passkeys can now be stored in your Bitwarden vault! Store and log in with passkeys using the Bitwarden browser extension (see [here](https://bitwarden.com/help/storing-passkeys/)).

**Comment by paour** [comment]
I updated today (docker ~~`latest`~~ `testing` tag), and passkeys seem to be functional 🎉. I was able to enroll one on the [yubico playground](https://demo.yubico.com/playground) with the browser plugin, and see the passkey in my Vault.

![image](https://github.com/dani-garcia/vaultwarden/assets/90963/a1ffb537-59bd-42d7-9bfc-1caab46eb898)

 I tried Firefox, but the plugin wasn't yet updated to version 23.10.1. Chrome's plugin was updated, and that worked.

**Reply by sky96111** [reply]
umm... That's strange.
My ``browser extension 2023.10.1`` doesn't ask me to create or use a passkey with ``Vaultwarden 1.29.2`` backend, but official ``bitwarden/self-host:2023.10.1-beta`` backend works like a charm

What did you do to make it work on vw?

**Reply by paour** [reply]
You need to be running the ~~`latest`~~ `testing` (unreleased) version. Either from source or docker.

**Reply by BlackDex** [reply]
You need Vaultwarden testing tagged images

**Reply by sky96111** [reply]
Great! It works! Thanks

**Reply by archef2000** [reply]
But why is chrome updated but not firefox it is "the same code"?

**Reply by BlackDex** [reply]
Because Mozilla needs to audit it first before they release a new version.

**Reply by archef2000** [reply]
Oh yeah right and we all know chrome aint doing that /s

**Comment by itsthejb** [comment]
Working here also (Yubico playground)

Now I just don’t understand other things! For example, sign into AppleId - how do I save a passkey for an existing account? I can try to sign in with one, but I’ve never created one. Can anyone help?

**Comment by archef2000** [comment]
Look under https://appleid.apple.com/account/manage to set one up

**Reply by itsthejb** [reply]
Yeah I’ve looked. Cannot see anywhere is there about creating a passkey

**Reply by archef2000** [reply]
Then account security and then add security key

**Reply by itsthejb** [reply]
Tried that also. This sends a notification to my phone, which redirects to the settings page. From there you can follow `Security Keys: Set Up >`, but this flow is entirely about setting up physical keys. I understand that passkeys are a related tech, but it's asking me to have 2 keys to physically tap on the device.

Sorry to be annoying! As far as I can see, I've got everything set up and the Yubico example worked fine

**Reply by archef2000** [reply]
Sorry just now that apple is apple

**Reply by itsthejb** [reply]
Haha. Well I wouldn't mind, but login on the website defaults to preferring a passkey, but I don't have one set up... 😑

**Reply by archef2000** [reply]
Under that button is a text that it only supports IOS > 16 so only with a phone

**Reply by itsthejb** [reply]
Already running iOS 17! No idea why I can't do it. Just been adding passkeys to some sites such as Google, Github, Amazon. Working smoothly there 🤷

**Reply by BlackDex** [reply]
Um, i think mobile support for passkeys isn't yet in there apps, only the browser extensions

**Reply by itsthejb** [reply]
Looking at https://www.passkeys.io/who-supports-passkeys, it suggests Apple/iCloud support may not be quite there yet. Which is surprising, because they've been pushing the feature... Could be it's only supported with iCloud keyring, which would be s**t

**Reply by agelwarg** [reply]
I'm not sure if this is correct but when storing passkeys in iCloud keychain, for everything OTHER THAN an Apple ID, you can actually see them stored in the password settings on both MacOS and iOS (synchronized).  However, when storing a passkey for your Apple ID, it seems to store that in a different location that we don't see (i.e., it's not listed under the saved Passkeys in iCloud).  Maybe it's only stored on device and not sync'd?  That might make sense since you need your Passkey to get into your Apple ID (iCloud, etc.).  I imagine that difference is somehow related to why we can't create a Passkey for an Apple ID in Bitwarden / Vaultwarden instead.  I'm not sure what's going on, but it does seem something is different for Passkeys for an Apple ID than other site(s).

**Comment by gkl1368** [comment]


**Reply by itsthejb** [reply]
Make sure you have all apps and extensions updated

Not sure if it's necessary, but this morning I also realised I didn't have websockets & push notifications properly set up. Possibly you need those to be correctly configured as well

**Reply by BlackDex** [reply]
It's only available in the browser extensions.
Nothing special in the web-vault.

**Reply by agelwarg** [reply]
It looks like we can't create/save new Passkeys using the (iOS) mobile app, but is it that we can't use them on mobile browsers (i.e., Safari, etc.) either?

**Reply by itsthejb** [reply]
I only saved them using the Chrome extension so far, yeah. But as far as I'm aware the support is in the apps. However, I wonder if there's really a flow that works that way?

**Reply by idoodler** [reply]
I just updated everything.

- iOS prompts me to define how Passkeys are stored (I selected Bitwarden)
- Then I go back, retry it and I am prompted with the Passkey dialogue and a continue button which does nothing.

I guess the iOS App is not ready yet.

**Reply by agelwarg** [reply]
I tried with upstream server (Bitwarden) itself as I keep an account there just for testing and got the same thing, so it definitely doesn't look ready on mobile yet.

**Reply by ilbarone87** [reply]
How do you guys are testing this if the server version has not been bumped yet on vaultwarden?

**Reply by paour** [reply]
See this comment https://github.com/dani-garcia/vaultwarden/discussions/3355#discussioncomment-7475167

**Reply by ilbarone87** [reply]
> See this comment https://github.com/dani-garcia/vaultwarden/discussions/3355#discussioncomment-7475167

Yeah, just saw it. Thanks!

**Reply by zvrh** [reply]
> Q: Can I store passkeys in the mobile app?
A: Passkeys support for mobile applications is planned for a future release.
by https://bitwarden.com/help/storing-passkeys/

**Reply by agelwarg** [reply]
@zvrh I saw that, but it wasn't clear if you could use the passkeys (already) stored in the vault, and it does not look to me like you can -- so there's basically no support on mobile at all yet.

**Reply by itsthejb** [reply]
I think this explains why I couldn’t save an Apple ID passkey as well

**Reply by leo15dev** [reply]
https://github.com/bitwarden/mobile/releases
Version 2023.10.0
iOS

View and edit login items with passkeys saved to them

https://github.com/bitwarden/clients/releases/tag/browser-v2023.10.0

Add support for saving and using passkeys to log in to apps and services

View and edit login items with passkeys saved to them

**Reply by agelwarg** [reply]
@leo15dev  I don't get it, do you?  I have that iOS version installed, and although you can see a Passkey in a Login item (i.e., "Created 11/4/23, 12:19PM") that was created from a desktop browser extension, you certainly can't create one on mobile, delete it on mobile, or even use it on mobile.

**Reply by leo15dev** [reply]
“Add support for saving and using passkeys to log in to apps and services”

We need wait the Bitwarden team to add this feature to the mobile app like they already added in the browser extension . Otherwise, we can not save or use passkeys in the mobile app.

**Reply by agelwarg** [reply]
I agree, but their release note is misleading

**Reply by tonyxu-io** [reply]
Clear to me. It said you can only view and edit. 

Especially in support page: https://bitwarden.com/help/storing-passkeys/#passkey-storage

> Saving and using passkeys are a feature of the Bitwarden browser extension. Other Bitwarden clients can be used to view the saved passkey.

**Reply by itsthejb** [reply]
“ Passkeys support for mobile applications is planned for a future release.”

so it’s coming

**Reply by agelwarg** [reply]
I suppose so, fair enough.  Regardless, the future is bright.

**Reply by TheBigBear** [reply]
Is this out yet for mobile iOS clients ? So we can start using passkeys across all devices without being tied to the Apple systems?

**Comment by archef2000** [comment]
When will stable be released?

**Reply by BlackDex** [reply]
Vaultwarden doesn't have any release schedules or something like that. So there is no eta of a new stable version.

Though we probably release one in the 1.29.x range soon.

**Reply by archef2000** [reply]
Will passkeys be in it?

**Reply by BlackDex** [reply]
It will be what currently is in testing, so yes

**Reply by archef2000** [reply]
Good thx

**Comment by RazielKanos** [comment]
Can you also use passkeys to log into vaultwarden? Currently my interface at "security" in my settings doesn't show anything about passkeys

**Comment by leo15dev** [comment]
@RazielKanos https://bitwarden.com/passwordless-passkeys/ 

"Sign into Bitwarden with passkeys
Coming soon, use passkeys to sign into Bitwarden to streamline authentication and use passkeys for vault encryption."

Just wait for Bitwarden to release this feature first.

**Reply by BlackDex** [reply]
I can sign in with passkey as a Webauthn 2fa. Which is fine by me. Fully passwordless could be nice though.

**Reply by AlexCherrypi** [reply]
Is this feature (fully passwordless login) planned? I would really appreciate it ...

**Comment by tonyxu-io** [comment]
Looks like many people got it to work, but I can't seem to get the Bitwarden passkey prompt shown even with testing tag. My Vaultwarden version shows 2023.10.0 and Chrome extension version 2023.10.2.

**Reply by itsthejb** [reply]
I believe you need to have websockets set up correctly for it to work. But someone can confirm. Double check your configuration

**Reply by BlackDex** [reply]
Websockets are not needed for passkeys

**Reply by ilbarone87** [reply]
> I believe you need to have websockets set up correctly for it to work. But someone can confirm. Double check your configuration

Websocket will be deprecated from next release so nothing to do with passkeys

**Reply by itsthejb** [reply]
Websockets aren’t being deprecated. What’s changing is that they work through the main port, ie not an extra port 3012. So some people need to update their config

**Reply by archef2000** [reply]
Have you relogedinto your vault from the extention?

**Comment by BobWs** [comment]
> Looks like many people got it to work, but I can't seem to get the Bitwarden passkey prompt shown even with testing tag. My Vaultwarden version shows 2023.10.0 and Chrome extension version 2023.10.2.

I'm having the same problem, passkey doesn't showup in my FireFox extension! Do I need to configure something else I don't know about? I see Github supports passkey (if it is the same feature) and would like to try that with Vaultwarden but I don't see the ability in my vaultwarden extension.
![Screen_Shot 1](https://github.com/dani-garcia/vaultwarden/assets/5485820/e08191a2-ae42-41d4-a212-067f0f68e0b1)
![Screen_Shot](https://github.com/dani-garcia/vaultwarden/assets/5485820/c9693e91-9b0e-44d8-920f-231488a798c2)
![Screen_Shot 2](https://github.com/dani-garcia/vaultwarden/assets/5485820/78879bec-e9fb-402f-8f0c-5452543be934)

**Reply by BlackDex** [reply]
Have you created a passkey for your GitHub account already? If not, then that would be the first step of course.

**Reply by BobWs** [reply]
When I try to add a passkey there is no popup from vaultwarden extension nor any ability to safe the passkey... 
![Screen_Shot 3](https://github.com/dani-garcia/vaultwarden/assets/5485820/45be8332-f525-48ae-876b-82f3fd8eb3ba)

**Reply by archef2000** [reply]
Have you updated the container?

**Reply by RazielKanos** [reply]
your server version is outdated

**Reply by BlackDex** [reply]
Server version looks ok too me

**Reply by leo15dev** [reply]
@BobWs 
Just cofirmed that

Windows 11 23H2
Firefox 119.0.1
Bitwarden Firefox Extension 10.2
Vaultwarden 1.30
 
Can login github with passkeys which is stored in Vaultwarden before.

Your firefox version is 115 esr vision, maybe it would not support passkey? 
Ref https://community.bitwarden.com/t/passkeys-and-firefox/58042

**Reply by Glandos** [reply]
https://caniuse.com/?search=passkey

It strangely says that no Firefox version support it.

**Reply by BlackDex** [reply]
I'm currently on Firefox 120, but it worked also on 119 for me on Linux, no issues at all.
I also tested Firefox ESR 115, and that also works just fine.

**Reply by kpfleming** [reply]
> https://caniuse.com/?search=passkey
> 
> It strangely says that no Firefox version support it.

It says that because Firefox on Linux does not have *native* support for passkeys, unlike Windows and macOS. Installing the Bitwarden extension adds support for passkeys.

**Reply by BlackDex** [reply]
Passkeys as in YubiKey/Webauthn support? Because that works just fine.

**Reply by BobWs** [reply]
reinstalled the FF add-on but still not working!

**Reply by BlackDex** [reply]
Try https://www.passkeys.io/ Works just fine for me.

**Reply by BobWs** [reply]
Tried  https://www.passkeys.io/ and it is working with VW FireFox extentions on a Windows 11 machine but not on my MacBook Pro with macOS 10.13 (High Sierra).
But the github passkey doesn't work on Windows, when I choose use passkey then I get a windows popup windows for an USB device.
On my iPad Github passkey works!

Any idea how to get Github passkey to work with Bitwarden extension?

**Reply by idoodler** [reply]
Github Passkeys work fine for me on macOS, i am using a chromium based browser tho (Arc).

**Comment by laurentftech** [comment]
Working  for me on MacOS with Safari.
The only issue is that the passkey window can take 15s or more to show up. Vaultwarden is hosted on a Synology 420+ on my local network. I am connected with LAN.  I’ve tested the same website with a passkey stored on Bitwarden.com and it show up in 2s.

**Reply by t3hk0d3** [reply]
@Gandulf78 AFAIK its not Vaultwarden's issue, because all secrets are synced locally. Vaultwarden is just an remote centralized storage.

**Reply by laurentftech** [reply]
> @Gandulf78 AFAIK its not Vaultwarden's issue, because all secrets are synced locally. Vaultwarden is just an remote centralized storage.

Ok. So it would be a browser extension performance issue with passkeys? Thanks.

**Reply by t3hk0d3** [reply]
@Gandulf78 AFAIK yes. Passkey support in Bitwarden's extensions are quite immature, and doesn't work ideally yet. For example for me it doesn't work reliably in Ungoogled Chrome.

**Reply by Glandos** [reply]
I had to log out and log in again in the extension before seeing the passkey popup from browser extension.

**Comment by ashoktvm** [comment]
Tried the latest version, the testing version, and specifically selected the 1.30.1 version, but none of them shows the passkey option in webvault or in extension what to do

**Reply by BlackDex** [reply]
There is no passkey option or something similar.
With the correct extension version installed and an unlocked extension you should be able to create a passkey and the extension will pop-up.

As suggest above try it on https://www.passkeys.io/

**Reply by ashoktvm** [reply]
Got it. It worked. But not working for google.

**Reply by mlcsthor** [reply]
A [temporary fix](https://github.com/bitwarden/clients/pull/6847) for the extension  will soon be deployed for passkeys that don't work on Google

**Comment by ilbarone87** [comment]
Does anyone know if there's a PR to track the development of the mobile client feature for passkeys?

**Comment by B0rner** [comment]
I read all the entries, but at the end I'm confused a bit, so sorry for asking again: is there currently a way for passwortless login to vaultwarden using a passkay (from Yubikey oder Google Titan) by using the vaultwarden web-interface or the bitwarden app?

**Reply by archef2000** [reply]
It is planned but not yet implemented:
https://bitwarden.com/de-DE/blog/bitwarden-passkey-management/

**Comment by MaximilianArnold** [comment]
Seems like password less sign in is now in beta
https://bitwarden.com/help/login-with-passkeys/

**Reply by ashoktvm** [reply]
I don't see the benefit at all. Since its aksing for master password again. Then whats the point even?

**Reply by MaximilianArnold** [reply]
Then you are doing something wrong. "Passkeys can be used to log in to Bitwarden as an alternative to using your master password and email."

**Reply by ashoktvm** [reply]
I did exactly as mentioned. It verifies my passkey and afterwards it asks for master password

**Reply by ashoktvm** [reply]
This is the message i get after using passkey

![image](https://github.com/dani-garcia/vaultwarden/assets/45116583/85d71b18-25b5-4031-9da1-4624beef5ab4)

**Reply by MaximilianArnold** [reply]
I really think this feature is not what you are looking for. The password less is not implemented in vaultwarden yet. Only login with another device who has a vault unlocked

**Reply by ashoktvm** [reply]
no this is correct. But the issue is bitwarden doesnt accept windows hello pin as passkey.

![image](https://github.com/dani-garcia/vaultwarden/assets/45116583/6812304a-a54c-4b32-9b1d-d88aedfe4ffa)

**Reply by ilbarone87** [reply]
https://github.com/bitwarden/mobile/pull/3044

Finally coming to mobile too. Hope won't take too long to finish this PR

**Reply by leo15dev** [reply]
@ilbarone87  The bitwarden mobile app in iOS TestFlight now already support passkey login (work with vaultwarden 1.30.5 ), so just wait for the official release.

**Reply by ilbarone87** [reply]
Yeah, I'm already using it. Seems alright. I only have one issue where in certain situations the app stop recognising the passkey.

**Reply by oguzhane** [reply]
was there any progress in vaultwarden to support login with passkeys?

**Reply by ilbarone87** [reply]
Has been implemented in main branch and is fully working.

**Reply by Gerardv514** [reply]
Logging into Vaultwarden using passkey is still not supported. 

Using passkeys in login items is and has been working

**Comment by freekvh** [comment]
Beware, I just locked myself out of an account because I saw the option "use passkey", went for it, it all worked but then somehow the key disappeared. Must be something to do with the official browser plugin in and lacking support if Vaultwarden? No idea. I should have investigated first...

**Reply by MaximilianArnold** [reply]
What you mean? You always have the option to use the master key? Passkey ist just more convenient

**Reply by BlackDex** [reply]
Maybe the site @freekvh used does not allow anything else, which would be strange. But normally there is a recovery feature via recovery keys.

**Reply by danieletorelli** [reply]
Which website it was? So maybe somebody else can try it

**Comment by danieletorelli** [comment]
I'm having some issue with passkeys in Safari (latest version). If I try to click on "Sign in with passkey" button on passkeys.io the action doesn't open Bitwarden extension. Simply it does nothing.

In the console I can see the warning:
```
User gesture is not detected. To use the WebAuthn API, call 'navigator.credentials.create' or 'navigator.credentials.get' within user activated events.
```

Does anyone know more about this?

**Reply by ericmail84** [reply]
It works for me. Have you signed up with passkeys.io already? E.g. you have signed up and created a passkey?

**Reply by danieletorelli** [reply]
Yes, I signed up. I suspect Adguard is interfering with the process. Disabling it makes it work, not always, but most of the times.

**Comment by Gerardv514** [comment]
If logins start moving toward passkeys, which is a way to eliminate passwords, what do you do when you jump on a shared or public computer and need to login? From my understanding of passkeys, and I haven’t started to use them yet since I’m waiting for it to mature a bit, you need BW browser extension to pass the passkey; this is fine if it’s your own personal device but what if it’s not? Or do you have to revert to the password in this scenario?

**Reply by kpfleming** [reply]
That's correct; for a login that has *only* a passkey, you won't be able to log in using a device (computer, mobile, etc.) unless it has the passkey available. If the site offers other login methods (like "forgot my password?" you may be able to use those.

**Reply by Gerardv514** [reply]
Yikes! That sucks. That will definitely be a short fall with this. At my job we utilize cars that have computers in them, they have a frozen image so at restart any setup you do needs to be done again. Usually I just add the BW browser extension, but the time spent in a car is usually short and this is tedious to setup when jumping in random cars every couple of days.

I wish they would implement some sort of way to send a push notification to the BW mobile client.

**Reply by denwald** [reply]
There is a solution for cross-device authentication using Passkeys. It's called _hybrid transport_ or _FIDO Cross-Device Authentication_. It should be supported on most recent operating systems already. I have no knowledge if third-party clients like Bitwarden will be able to support it too. There is one catch, though. The device where you want to authenticate using the cross-device authentication needs a working Bluetooth Low Energy transmitter and a network connection.

**Comment by AlphaCraft9658** [comment]
I am confused. Has the support for storing passkeys already been implemented in Vaultwarden or not?

**Reply by ilbarone87** [reply]
Not for the mobile application, only actually in beta

**Reply by BlackDex** [reply]
Vaultwarden supports storing the keys. Only not all clients support this yet.

**Reply by AlphaCraft9658** [reply]
> Vaultwarden supports storing the keys. Only not all clients support this yet.

Alright. I tested the functionality with webauthn.io and it worksy but only.with the browser extension. Is the beta with this feature the latest Google Play beta?

**Reply by ilbarone87** [reply]
https://community.bitwarden.com/t/about-the-beta-program/39185

Have a look here. 
Also if anyone else is doing the beta for iOS I found the app crashing quite often and also not able to authenticate with passkey anymore after a while until I don't log out and go back in. Let me know if anyone is experiencing the same behaviours

**Reply by Gerardv514** [reply]
Yes I’ve seen the beta app crash as well, every time I added a TOTP from the browser extension.

**Reply by leo15dev** [reply]
2024.4.2
Use passkeys on mobile apps: Password Manager mobile apps can now be used to create and sign in with passkeys. This feature is available for iOS and as a beta for Android. Learn more here.

https://bitwarden.com/help/releasenotes/

**Reply by eddieirvine** [reply]
I’m running Vaultwarden 2024.1.2 and the iOS app 2024.4.2 (7528).

I don’t see anything related to passkeys…

**Reply by AlphaCraft9658** [reply]
> I’m running Vaultwarden 2024.1.2 and the iOS app 2024.4.2 (7528).
> 
> I don’t see anything related to passkeys…

It seems like one can't manually add passkeys. Whenever a website requests a passkey for login, the browser extension asks you to store or use the passkey, and once stored from there, you can see it in the web interface. The mobile apps don't yet have support for it. Well, the iOS app seems to have it, but on Android, it's still in beta.

**Reply by eddieirvine** [reply]
> It seems like one can't manually add passkeys. Whenever a website requests a passkey for login, the browser extension asks you to store or use the passkey, and once stored from there, you can see it in the web interface. The mobile apps don't yet have support for it. Well, the iOS app seems to have it, but on Android, it's still in beta.

Thanks.
You're right, via the desktop browser extension it works!
If a passkey is then stored, it also pops up on the iPhone when logging to a website.

**Comment by Timtam** [comment]
Thought I'd post here since i've got one specific use-case where passkeys with the current Bitwarden chrome extension doesn't seem to work right, and I don't know if either the web app, the extension or Vaultwarden is at fault here. I'm talking about Miniflux v2 (https://github.com/miniflux/v2). I created a passkey and stored it in Vaultwarden, that part worked just fine. Whenever I try to login, Bitwarden tells me that it couldn't find a passkey and prompts me to either use local auth or cancel. Passkeys work fine in any other scenario though. Do you have any idea why that might be? Is it Miniflux related and I should open an issue over there?
Thanks!

**Reply by AlphaCraft9658** [reply]
> Thought I'd post here since i've got one specific use-case where passkeys with the current Bitwarden chrome extension doesn't seem to work right, and I don't know if either the web app, the extension or Vaultwarden is at fault here. I'm talking about Miniflux v2 (https://github.com/miniflux/v2). I created a passkey and stored it in Vaultwarden, that part worked just fine. Whenever I try to login, Bitwarden tells me that it couldn't find a passkey and prompts me to either use local auth or cancel. Passkeys work fine in any other scenario though. Do you have any idea why that might be? Is it Miniflux related and I should open an issue over there?
> Thanks!

Could it potentially be about a potentially different subdomain? Maybe it uses a slightly different domain or path for login, which Bitwardn/Vaultwarden doesn't recognize, as a different login page has been stored for the passkey. This is my suspicion, but also my only idea as to what could be going on for you.

**Reply by Timtam** [reply]
I don't know really. I just tried again, deleted the old passkey from Vaultwarden and Miniflux and added a new one. Tried to login and it just says "no passkey could be found". I checked my Miniflux config and it got the correct base url to work with, but I don't know what the web authenticator actually sends as a url to Bitwarden/Vaultwarden that cannot be found.

**Reply by ilbarone87** [reply]
There are other cases, I have same issue with Teleport Community Edition. Same behaviour. Passkey in vaultwarden doesn't seem to be recognised. Tried to delete and create. Creation works well but login doesn't work.

**Reply by DaStivi** [reply]
i would also confirm there must be an issue... 
in my case i'll test with M365 Passkey support... its even more complicated there, because you've to set allowed passkey provider AAIDs ... i've entered the Bitwarden AAID there in the assumption the "client" or browser extension is providing this ID... 

sadly there is no error or log message that would help only some non-sense message..
there is passkey created successfull message.. then you can give it  ah name but afterwards it fails to save it to the account..

![image](https://github.com/dani-garcia/vaultwarden/assets/7464802/59e690fc-ca41-4420-8ab7-0e93b1b89661)

passkey also created in bit/vault-warden obviously..
![image](https://github.com/dani-garcia/vaultwarden/assets/7464802/4787e010-63d1-4c04-808f-201ac820af1c)

**Comment by TheBigBear** [comment]
No, there is no issue at all. Microsoft currently do not allow passkeys from any passkeys syncing accounts. only what they call device bound passkeys are supported. 

so you have to either store your passkey in the microsoft authenticator app and allow that app to be the password provider and that will work. 

but trying to store the passkey in the apple iCloud Keychain or a password manager like bitwarden, vaultwarden or keepassxc will fail, by MS design. 

HTH

**Reply by DaStivi** [reply]
> No, there is no issue at all. Microsoft currently do not allow passkeys from any passkeys syncing accounts. only what they call device bound passkeys are supported. so you have to either store your passkey in the microsoft authenticator app and allow that app to be the password provider and that will work. but trying to store the passkey in the apple iCloud Keychain or a password manager like bitwarden, vaultwarden or keepassxc will fail, by MS design. HTH
> […](#)
> On 7. Jun 2024, at 11:14, DaStivi ***@***.***> wrote: i would also confirm there must be an issue... in my case i'll test with M365 Passkey support... its even more complicated there, because you've to set allowed passkey provider AAIDs ... i've entered the Bitwarden AAID there in the assumption the "client" or browser extension is providing this ID... sadly there is no error or log message that would help only some non-sense message.. there is passkey created successfull message.. then you can give it ah name but afterwards it fails to save it to the account.. image.png (view on web) <https://github.com/dani-garcia/vaultwarden/assets/7464802/59e690fc-ca41-4420-8ab7-0e93b1b89661> passkey also created in bit/vault-warden obviously.. image.png (view on web) <https://github.com/dani-garcia/vaultwarden/assets/7464802/4787e010-63d1-4c04-808f-201ac820af1c> — Reply to this email directly, view it on GitHub <[#3355 (reply in thread)](https://github.com/dani-garcia/vaultwarden/discussions/3355#discussioncomment-9699600)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AADTAQKQS2KY5OZSHOIVVDDZGF2WVAVCNFSM6AAAAAAWAPKML6VHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4TMOJZGYYDA>. You are receiving this because you commented.

thx for that info.. didn't know about it!

**Comment by aikooo7** [comment]
Am I the unique who still can't passkeys in vaultwarden in bitwarden client in android? Is it a bitwarden client problem or vaultwarden problem?

**Reply by TheBigBear** [reply]
@aikooo7 what passkey on what system ? Any passkey todo with O365 or M365? Or a Google or GitHub passkey?
Vaultwarden works with passkeys where it can. But the MS o365 and m365 passkeys are only allowed and working for device bound passkeys, which rules out vaultwarden/bitwarden.
Suggest you try and test or debug your passkey vaultwarden/Bitwarden functionality on a passkeys poc test site.
Without more detail about the issue you are facing we can’t really help or advise you any better.

**Reply by BlackDex** [reply]
You need at least Android 14 and i think you even need Google Android, not a forked Android.
And not all applications support it. For example FireFox supports it only in there current latest version if I'm correct.

**Reply by kpfleming** [reply]
I've got Android 14 (not a fork), Firefox 128, and the latest Bitwarden client, and even though I've told the system to use passkeys from Bitwarden they are not available in Firefox. This functionality is still not fully baked, for sure.

**Reply by BlackDex** [reply]
Could be. It is at least not something Vaultwarden can fix. Vaultwarden supports storing Passkeys.
The clients and platforms need to support using it.

**Reply by aikooo7** [reply]
> @aikooo7 what passkey on what system ? Any passkey todo with O365 or M365? Or a Google or GitHub passkey?
> Vaultwarden works with passkeys where it can. But the MS o365 and m365 passkeys are only allowed and working for device bound passkeys, which rules out vaultwarden/bitwarden.
> Suggest you try and test or debug your passkey vaultwarden/Bitwarden functionality on a passkeys poc test site.
> Without more detail about the issue you are facing we can’t really help or advise you any better.

I tried with passkeys.io

**Reply by TheBigBear** [reply]
@aikooo7 
> I tried with passkeys.io

I guess you don't want or need help? Second posting with no background or error message and zero detail of what it is that fails for you?

**Reply by aikooo7** [reply]
> @aikooo7 
> > I tried with passkeys.io
> 
> I guess you don't want or need help? Second posting with no background or error message and zero detail of what it is that fails for you? 
> 

Why would I post if I don't want help? How would I post an error message if I don't get any? O

I just don't get the prompt to save with bitwarden.

**Reply by Gerardv514** [reply]
@aikooo7 works for me

**Reply by aikooo7** [reply]
> @aikooo7 works for me

What browser you are using? I tried firefox and chrome, did you change anything in the browser?

I am also using the latest beta version of bitwarden client in android and latest vaultwarden in docker.

**Reply by Gerardv514** [reply]
> > @aikooo7 works for me
> 
> What browser you are using? I tried firefox and chrome, did you change anything in the browser?
> 
> I am also using the latest beta version of bitwarden client in android and latest vaultwarden in docker.

Chrome on iOS. 

Site:
![IMG_7265](https://github.com/user-attachments/assets/d5a009e2-7946-4189-924e-c65c6d55943a)

Bitwarden app:
![IMG_7266](https://github.com/user-attachments/assets/dcfd3c4f-b7ea-4dfe-86f8-67fc6e0beeba)
